Simplify logic and use one-time binding syntax in admin page

526avijitgupta opened the pull request:
Tasklist:
- [x] Found out that the logic in admin page was unnecessarily complex, and the same function was being called twice when switching tabs (due to it being bound to ng-click, as well as a watcher on url's hash!) -- thus simplified the logic.


- [x] ~~Use one-time binding for variables used as constants, or any other expressions.~~
- [x] ~~Get variables unbounded to `$scope` or `$rootScope` if not necessary.~~

^To be done in a new PR.

seanlip commented on core/templates/dev/head/admin/Admin.js:
```diff
@@ -16,64 +16,40 @@
+      if (newHash !== oldHash) {
+        for (var url in $scope.ADMIN_TAB_URLS) {
+          if ($scope.ADMIN_TAB_URLS[url] === newHash) {
+            $scope.currentTab = newHash;
```
then break? (this is minor)

526avijitgupta replied:
Done.

seanlip commented on core/templates/dev/head/admin/admin.html:
```diff
@@ -29,15 +29,18 @@
             <li>
               <span class="oppia-navbar-breadcrumb-separator"></span>
-              Admin
+                Admin
```
Shouldn't be indented, I think

526avijitgupta replied:
Ah, right. Reverted this back.

seanlip commented:
Excellent, this looks great -- just a couple minor things. Thanks @526avijitgupta!

Wasn't quite sure what the second/third items on your tasklist meant, though. Are those for future PRs?

526avijitgupta commented:
@seanlip Addressed all review comments. PTAL.
* In the latest commit, I have gone ahead and replaced `$scope.adminHandlerUrl` by `var adminHandlerUrl` since I could not see any reason to keep this variable on `$scope`. On a similar note, I guess there would be instances in code where variables might be bound to `$scope` when they can actually be directly used instead. So I'll go ahead and fix those as well.

* I had opened this PR with the intent to use one-time binding syntax in all possible places, but actually ended up refactoring the code for admin page first.
I would lean opening a new PR to work on one-time binding and also making variables which are bound to $scope to be used directly (as pointed out above). wdyt ?

seanlip commented:
Thanks @526avijitgupta! What you say in the first bullet point sounds good. Also, making a new PR sounds good too. Thanks!

seanlip commented on core/templates/dev/head/admin/Admin.js:
```diff
@@ -16,64 +16,41 @@
+  function($scope, $http, UrlInterpolationService, ADMIN_TAB_URLS) {
     var ADMIN_JOB_OUTPUT_URL_PREFIX = '/adminjoboutput';
-    $scope.adminTopicsCsvDownloadHandlerUrl = '/admintopicscsvdownloadhandler';
+    var adminHandlerUrl = '/adminhandler';
```
Worth making these two things constants (e.g. ADMIN_HANDLER_URL), for consistency?

526avijitgupta replied:
Yes, done.

seanlip commented:
Also, made one comment. Not sure why e2e tests are failing, but I suspect it is to do with this change and that running the problematic tests locally will illustrate what the issue is.

seanlip commented:
Hm, e2e tests still seem unhappy. Any idea why?

MAKOSCAFEE commented:
@526avijitgupta  ping?

526avijitgupta commented:
@MAKOSCAFEE Will be taking a look at this latest by the end of the coming week. Busy with college related work currently.

seanlip commented:
Hi @526avijitgupta -- looking at this again; the e2e tests still fail. Did you figure out why?

Thanks!

526avijitgupta commented:
@seanlip Unfortunately no. I've had a look at the output of the failing tests a couple of times, but haven't been able to conclude anything fruitful yet. Currently, I have my exams going on and I plan to sit down and fix this just after the exams are over (which should be by 13th/14th Sept).

seanlip commented:
Ah, ok -- sounds good. Thanks, and good luck with your exams!

seanlip commented:
Hi @526avijitgupta, just going through some old PRs. Should we wrap this one up?

526avijitgupta commented:
@seanlip Yeah, will make sure to work on this once #2296 gets in!

526avijitgupta commented:
@seanlip I've narrowed down the cause of failure of the e2e tests to a single function in `protractor_utils/admin.js`: https://github.com/oppia/oppia/blob/develop/core/tests/protractor_utils/admin.js#L30-L57 -- but am still in the process of figuring the reason for the same!

526avijitgupta commented:
@seanlip Please review.

seanlip commented:
Just one thing. Bad merge?

526avijitgupta commented on core/templates/dev/head/pages/admin/admin.html:
```diff
@@ -336,7 +399,28 @@
         </md-card>
 
-        <md-card class="oppia-page-card oppia-long-text" ng-if="currentTab === TAB_MISC" style="max-width: 700px">
+        <md-card class="oppia-page-card oppia-long-text"
```
You were right. Done.

526avijitgupta commented:
@seanlip PTAL

seanlip commented:
Not sure if I'm imagining it, but the admin tabs still feel more sluggish than before. @BenHenning @wxyxinyu could one of you PTAL?

seanlip commented:
(I have no code comments.)

seanlip commented:
Update: I just tested on Chrome, but I can't really see a difference between develop and this branch. So I'm not sure if I'm making it up.

I think all that's needed here is for someone else to check if there's a noticeable speed difference on the admin page (when switching tabs) between develop and this branch. If not, I think this is fine to merge.

wxyxinyu commented:
I tested the admin page. No noticeable time difference for me by eyeballing it, but didn't time it either.

BenHenning commented:
Sorry for not mentioning this earlier, but for context @526avijitgupta and I talked through this a bit during the bug bash. The sluggishness happened when anchors only changed their href and we tried responding to those changes. For some reason either the change wasn't fired right away or we were somehow blocking up the $digest cycle, but if you clicked on the Config tab in the admin panel and kept your mouse still immediately after clicking then the page wouldn't load until you moved your mouse. The sluggishness had to do with event propagation. To fix this @526avijitgupta reverted back to using ``ng-click`` to bind the events, like the admin page did before (and how most of Oppia does it).
